### ISSUE TITLE
Typed macros

### ISSUE THREAD

**Issue by axeld** [PebbleTemplates/pebble#406]
It would be great to have typed macros. For example, in one of my templates I need to detect empty values at several occasions. However, an empty value is not just 'empty', but also '-' is considered an empty value.

Now I have to do something like that:
`{% if hasValue(myValue) | trim == 'yes' %}... show my value ...{% endif %}`

Which is pretty hard on the eyes. If I could do the trim in the macro, that would already be a big improvement, but actually just writing:
`{% if hasValue(myValue) %}...{% endif %}`

Would be nice.

**Comment by jochen777** [comment]
You can write your own function (see here: https://github.com/PebbleTemplates/pebble/wiki/extending-pebble)

There you can perform whatever you want to check.

**Comment by ebussieres** [comment]
I'm surprised that the trim cannot be made in the macro. Can you provide the content of your "hasValue" macro ?

**Comment by axeld** [comment]
The macro looks like this:
 `{% macro hasValue(value) %}{% if value is empty or value == '-' %}no{% else %}yes{% endif %}{% endmacro %}`

I'm afraid that adding a function is not an option, as the application itself is a black box -- I have only access to the template.

**Comment by axeld** [comment]
Without the 'trim' the evaluation is never true.

**Comment by ebussieres** [comment]
It seems a bit weird to have a macro that return "yes" or "no". Macro should be used for content reusing, not defining a function.

Like @jochen777 said, it's a perfect use case for a custom function. See here: https://github.com/PebbleTemplates/pebble/wiki/extending-pebble

If you need help to define your custom function, you can ask here.

**Comment by DirkLachowski** [comment]
@axeld Maybe off topic, but that discussion feels like you are trying to "repair" your model in the view/template because you are handing over a model (unchecked) you get from a third party. Wouldn't you be better of providing a simple model to the template and keep complex logic were you are preparing the model?

**Comment by axeld** [comment]
Of course, it would be better, but I cannot change that. I have to put the logic in the template, as I don't have the means to add a function or fix the model.

I mean, all this if/for/... functionality is there for a reason; I cannot always fix the model, and templates may become very complex. I'm just asking to improve upon that foundation. AFAIK, TWIG does not have this functionality either, though.

**Comment by ebussieres** [comment]
IMHO, having complex templates is always a bad idea. It's hard to maintain and hard to test too.

But I'd like to understand why you don't have access to add a function or fix the model ? What's your current setup ?

**Comment by axeld** [comment]
I'm using a product information management system that has Pebble as an export channel. The data comes from the product management, and we use Pebble to create print labels (via HTML->PDF conversion).

In the template, I just either get an 'item' map variable that has all the product info attached, or a list of such items. I cannot alter the data beforehand, and I have no means to change the interaction with Pebble itself. All I can access is the template. While I could add extra variables to the template, this method won't work, as the template must be usable for a single item as well as a list of items (in which case the PDF will contain all labels at once).

Does that make sense to you?
